android/src/main/java/com/clevertap/react/CleverTapEventEmitter.kt

```diff
@@ -0,0 +1,122 @@
+package com.clevertap.react
+
+import android.util.Log
+import com.facebook.react.bridge.ReactContext
+import com.facebook.react.modules.core.DeviceEventManagerModule
+import java.util.LinkedList
+import java.util.Queue
+
+/**
+ * CleverTapEventEmitter is responsible for emitting events to the React Native JavaScript layer.
+ * It manages event buffers and allows events to be queued which is useful when needing to send an
+ * event while the ReactContext of the application is not yet initialized. Only events specified as
+ * [CleverTapEvent.bufferable] will be considered for buffering, all other events will be emitted
+ * immediately.
+ */
+object CleverTapEventEmitter {
+    private const val LOG_TAG = "CleverTapEventEmitter"
+
+    var reactContext: ReactContext? = null
+
+    private class Buffer(var enabled: Boolean) {
+        val items: Queue<Any?> by lazy { LinkedList() }
+    }
+
+    private var eventsBuffers: Map<CleverTapEvent, Buffer> = createBuffersMap(enableBuffers = true)
+
+    /**
+     * Enable buffering for the specified event. While enabled, all events which should be emitted
+     * will be added to the buffer. Only buffers for events specified as [CleverTapEvent.bufferable]
+     * can be enabled.
+     *
+     * @see [disableBuffer]
+     */
+    fun enableBuffer(event: CleverTapEvent) {
+        eventsBuffers[event]?.enabled = true
+    }
+
+    /**
+     * Disable buffering for the specified event. This will only have effect for events specified as
+     * [CleverTapEvent.bufferable].
+     *
+     * @see [enableBuffer]
+     */
+    fun disableBuffer(event: CleverTapEvent) {
+        eventsBuffers[event]?.enabled = false
+    }
+
+    /**
+     * Clear all the buffered events from all buffers and set whether all buffers should be enabled
+     * after that.
+     *
+     * @param enableBuffers enable/disable all buffers after they are cleared
+     */
+    fun resetAllBuffers(enableBuffers: Boolean) {
+        eventsBuffers = createBuffersMap(enableBuffers)
+        Log.i(LOG_TAG,"Buffers reset and enabled: $enableBuffers")
+    }
+
+    /**
+     * Emit all currently buffered events for the specified event name.
+     */
+    fun flushBuffer(event: CleverTapEvent) {
```

Comment:
Should the event be removed from eventBuffers once flushed?

Reply:
Yes, it is better to remove it as soon as possible. We've added those buffers for the very specific case when an event is raised before RN is loaded. RN events are fire-and-forget. We send the buffered events right away when a listener is attached (and we clear them after 5 secods) since those events would become stale (and potentially unexpected) if they are send much later. This is the same behavior as if an event is send, but there are no listeners added for it - then the event would be lost. 

I believe in the event of attaching the listener twice, the client should expect only the first listener to receive initial events.

Reply:
Yess. This would be incorrect integration from the client anyways but we can avoid NOCs by handling it in the SDK?
Directly sent events will always be received only on the latest attached listener and hence we can make things consistent?